File: go/dcf/frame.go
// SPDX-License-Identifier: LGPL-3.0-only

// Package dcf implements the DeMoD 17-byte DeModFrame wire quantum (version 1),
// byte-identical to the reference codec in python/MCP/wirelab_core.py and certified
// against Documentation/golden_vectors.json (the cross-language contract).
//
// Wire layout (big-endian):
//
//	[0]     sync = 0xD3
//	[1]     flags: version[7:4]=1 | frame_type[3:0]
//	[2:4]   seq      u16
//	[4:6]   src      u16
//	[6:8]   dst      u16
//	[8:12]  payload  4 bytes
//	[12:15] ts_us    u24
//	[15:17] CRC-16/CCITT-FALSE over bytes [0..14]
package dcf

import (
	"encoding/hex"
	"errors"
	"fmt"
)

// Wire constants.
const (
	SyncByte  = 0xD3
	Version   = 1
	FrameSize = 17
	crcCover  = 15
	Broadcast = 0xFFFF
)

// FrameType is the 4-bit frame-type nibble.
type FrameType uint8

const (
	FData   FrameType = 0
	FAck    FrameType = 1
	FBeacon FrameType = 2
	FCtrl   FrameType = 3
)

// Decode errors.
var (
	ErrBadLength  = errors.New("dcf: frame must be 17 bytes")
	ErrBadSync    = errors.New("dcf: bad sync byte")
	ErrBadVersion = errors.New("dcf: bad version nibble")
	ErrBadCRC     = errors.New("dcf: CRC mismatch")
)

// Crc16CCITT computes CRC-16/CCITT-FALSE (poly 0x1021, init 0xFFFF, no reflection,
// no xorout) over data.
func Crc16CCITT(data []byte) uint16 {
	crc := uint16(0xFFFF)
	for _, b := range data {
		crc ^= uint16(b) << 8
		for i := 0; i < 8; i++ {
			if crc&0x8000 != 0 {
				crc = (crc << 1) ^ 0x1021
			} else {
				crc <<= 1
			}
		}
	}
	return crc
}

// Frame is a decoded DeModFrame with host-order fields.
type Frame struct {
	Version uint8
	Type    FrameType
	Seq     uint16
	Src     uint16
	Dst     uint16
	Payload [4]byte
	TsUs    uint32 // 24-bit; the high byte is masked off on encode
}

// Encode serialises f into exactly 17 bytes, computing and appending the CRC.
func (f Frame) Encode() [FrameSize]byte {
	var b [FrameSize]byte
	b[0] = SyncByte
	b[1] = ((f.Version & 0x0F) << 4) | (byte(f.Type) & 0x0F)
	b[2] = byte(f.Seq >> 8)
	b[3] = byte(f.Seq)
	b[4] = byte(f.Src >> 8)
	b[5] = byte(f.Src)
	b[6] = byte(f.Dst >> 8)
	b[7] = byte(f.Dst)
	copy(b[8:12], f.Payload[:])
	b[12] = byte((f.TsUs >> 16) & 0xFF)
	b[13] = byte((f.TsUs >> 8) & 0xFF)
	b[14] = byte(f.TsUs & 0xFF)
	crc := Crc16CCITT(b[:crcCover])
	b[15] = byte(crc >> 8)
	b[16] = byte(crc)
	return b
}

// Syndrome returns the affine validity syndrome of a 17-byte word: the word's CRC is
// valid iff Syndrome == 0.
func Syndrome(word []byte) (uint16, error) {
	if len(word) != FrameSize {
		return 0, ErrBadLength
	}
	stored := uint16(word[15])<<8 | uint16(word[16])
	return Crc16CCITT(word[:crcCover]) ^ stored, nil
}

// Decode parses a 17-byte wire buffer, validating sync, version nibble, and CRC.
func Decode(word []byte) (Frame, error) {
	if len(word) != FrameSize {
		return Frame{}, ErrBadLength
	}
	if word[0] != SyncByte {
		return Frame{}, ErrBadSync
	}
	if word[1]>>4 != Version {
		return Frame{}, ErrBadVersion
	}
	if syn, _ := Syndrome(word); syn != 0 {
		return Frame{}, ErrBadCRC
	}
	return Frame{
		Version: word[1] >> 4,
		Type:    FrameType(word[1] & 0x0F),
		Seq:     uint16(word[2])<<8 | uint16(word[3]),
		Src:     uint16(word[4])<<8 | uint16(word[5]),
		Dst:     uint16(word[6])<<8 | uint16(word[7]),
		Payload: [4]byte{word[8], word[9], word[10], word[11]},
		TsUs:    uint32(word[12])<<16 | uint32(word[13])<<8 | uint32(word[14]),
	}, nil
}

// exampleFrameFull is the golden anchor frame (Ctrl, seq=0x1234, src=1, dst=broadcast,
// payload=DEADBEEF, ts=0xAB12CD) from Documentation/golden_vectors.json.
const exampleFrameFull = "d31312340001ffffdeadbeefab12cd24c0"

// init self-certifies the codec against the CRC and example-frame anchors on import,
// panicking (refuse to load) if the implementation has diverged from the reference.
func init() {
	if c := Crc16CCITT([]byte("123456789")); c != 0x29B1 {
		panic(fmt.Sprintf("dcf: CRC anchor diverged: CRC(\"123456789\")=0x%04X, want 0x29B1", c))
	}
	if c := Crc16CCITT(make([]byte, 15)); c != 0x4EC3 {
		panic(fmt.Sprintf("dcf: CRC anchor diverged: CRC(0^15)=0x%04X, want 0x4EC3", c))
	}
	ex := Frame{Version: 1, Type: FCtrl, Seq: 0x1234, Src: 1, Dst: Broadcast,
		Payload: [4]byte{0xDE, 0xAD, 0xBE, 0xEF}, TsUs: 0xAB12CD}
	got := ex.Encode()
	if hex.EncodeToString(got[:]) != exampleFrameFull {
		panic(fmt.Sprintf("dcf: exampleFrame anchor diverged: got %s, want %s",
			hex.EncodeToString(got[:]), exampleFrameFull))
	}
}

File: swift/Sources/DCFWire/Frame.swift
// SPDX-License-Identifier: LGPL-3.0-only

// The DeMoD 17-byte DeModFrame wire quantum (version 1), byte-identical to the
// reference codec in python/MCP/wirelab_core.py and certified against
// Documentation/golden_vectors.json (the cross-language contract).
//
// Wire layout (big-endian):
//   [0]     sync = 0xD3
//   [1]     flags: version[7:4]=1 | frame_type[3:0]
//   [2:4]   seq      u16
//   [4:6]   src      u16
//   [6:8]   dst      u16
//   [8:12]  payload  4 bytes
//   [12:15] ts_us    u24
//   [15:17] CRC-16/CCITT-FALSE over bytes [0..14]

public let SYNC: UInt8 = 0xD3
public let VERSION: UInt8 = 1
public let FRAME_SIZE: Int = 17
public let CRC_COVER: Int = 15
public let BROADCAST: UInt16 = 0xFFFF

/// CRC-16/CCITT-FALSE (poly 0x1021, init 0xFFFF, no reflection, no xorout).
public func crc16(_ data: [UInt8], _ len: Int) -> UInt16 {
    var crc = 0xFFFF
    for k in 0..<len {
        crc ^= Int(data[k]) << 8
        crc &= 0xFFFF
        for _ in 0..<8 {
            if crc & 0x8000 != 0 {
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            } else {
                crc = (crc << 1) & 0xFFFF
            }
        }
    }
    return UInt16(crc & 0xFFFF)
}

public enum FrameError: Error {
    case badLength, badSync, badVersion, badCrc
}

public struct Frame {
    public var version: UInt8 = 1     // 4-bit
    public var type: UInt8 = 0        // 4-bit (0=Data,1=Ack,2=Beacon,3=Ctrl)
    public var seq: UInt16 = 0
    public var src: UInt16 = 0
    public var dst: UInt16 = 0
    public var payload: [UInt8] = [0, 0, 0, 0]
    public var tsUs: UInt32 = 0       // u24

    public init() {}

    /// Serialise into exactly 17 bytes, computing and appending the CRC.
    public func encode() -> [UInt8] {
        var b = [UInt8](repeating: 0, count: FRAME_SIZE)
        b[0] = SYNC
        b[1] = UInt8(((Int(version) & 0x0F) << 4) | (Int(type) & 0x0F))
        b[2] = UInt8(seq >> 8); b[3] = UInt8(seq & 0x00FF)
        b[4] = UInt8(src >> 8); b[5] = UInt8(src & 0x00FF)
        b[6] = UInt8(dst >> 8); b[7] = UInt8(dst & 0x00FF)
        b[8] = payload[0]; b[9] = payload[1]; b[10] = payload[2]; b[11] = payload[3]
        b[12] = UInt8((tsUs >> 16) & 0xFF)
        b[13] = UInt8((tsUs >> 8) & 0xFF)
        b[14] = UInt8(tsUs & 0xFF)
        let crc = crc16(b, CRC_COVER)
        b[15] = UInt8(crc >> 8); b[16] = UInt8(crc & 0x00FF)
        return b
    }
}

/// Affine validity syndrome of a 17-byte word: CRC-valid iff this returns 0.
public func syndrome(_ w: [UInt8]) -> UInt16 {
    precondition(w.count == FRAME_SIZE, "need 17 bytes")
    let stored = (UInt16(w[15]) << 8) | UInt16(w[16])
    return crc16(w, CRC_COVER) ^ stored
}

/// Parse a 17-byte buffer, validating sync, version nibble, and CRC.
public func decode(_ w: [UInt8]) throws -> Frame {
    guard w.count == FRAME_SIZE else { throw FrameError.badLength }
    guard w[0] == SYNC else { throw FrameError.badSync }
    guard (w[1] >> 4) == VERSION else { throw FrameError.badVersion }
    guard syndrome(w) == 0 else { throw FrameError.badCrc }
    var f = Frame()
    f.version = w[1] >> 4
    f.type = w[1] & 0x0F
    f.seq = (UInt16(w[2]) << 8) | UInt16(w[3])
    f.src = (UInt16(w[4]) << 8) | UInt16(w[5])
    f.dst = (UInt16(w[6]) << 8) | UInt16(w[7])
    f.payload = [w[8], w[9], w[10], w[11]]
    f.tsUs = (UInt32(w[12]) << 16) | (UInt32(w[13]) << 8) | UInt32(w[14])
    return f
}

/// Verify the codec against the CRC + example-frame anchors; throws on divergence.
/// Swift libraries have no load hook, so call this once at startup; the cert test
/// invokes it.
public func selfCert() throws {
    guard crc16(Array("123456789".utf8), 9) == 0x29B1 else { throw FrameError.badCrc }
    guard crc16([UInt8](repeating: 0, count: 15), 15) == 0x4EC3 else { throw FrameError.badCrc }
    var ex = Frame()
    ex.version = 1; ex.type = 3; ex.seq = 0x1234; ex.src = 1; ex.dst = BROADCAST
    ex.payload = [0xDE, 0xAD, 0xBE, 0xEF]; ex.tsUs = 0xAB12CD
    let want: [UInt8] = [0xD3, 0x13, 0x12, 0x34, 0x00, 0x01, 0xFF, 0xFF,
                         0xDE, 0xAD, 0xBE, 0xEF, 0xAB, 0x12, 0xCD, 0x24, 0xC0]
    guard ex.encode() == want else { throw FrameError.badCrc }
}
